cpp/include/raft/core/device_resources_manager.hpp

```diff
@@ -0,0 +1,578 @@
+ * without actively managing streams or other CUDA-specific objects.
+ *
+ * To control the resources a consuming application will use, the
+ * resource manager provides setters for a variety of values. For
+ * instance, to ensure that no more than `N` CUDA streams are used per
+ * device, a consumer might call
+ * `raft::device_resources_manager::set_streams_per_device(N)`. Note that all of these
+ * setters must be used prior to retrieving the first `device_resources` from
+ * the manager. Setters invoked after this will log a warning but have no
+ * effect.
+ *
+ * After calling all desired setters, consumers can simply call
+ * `auto res = raft::device_resources_manager::get_resources();` to get a valid
+ * device_resources object for the current device based on previously-set
+ * parameters. Importantly, calling `get_resources()` again from the same
+ * thread is guaranteed to return a `device_resources` object with the same
+ * underlying CUDA stream and (if a non-zero number of stream pools has been
+ * requested) stream pool.
+ *
+ * Typical usage might look something like the following:
+ * @code
+ * void initialize_application() {
+ *   raft::device_resources_manager::set_streams_per_device(16);
+ * }
+ *
+ * void foo_called_from_multiple_threads() {
+ *   auto res = raft::device_resources_manager::get_device_resources();
+ *   // Call RAFT function using res
+ *   raft::device_resources_manager::synchronize_work_from_this_thread();
+ * }
+ * @endcode
+ *
+ * Note that all public methods of the `device_resources_manager` are thread-safe,
+ * but the manager is designed to minimize locking required for
+ * retrieving `device_resources` objects. Each thread must acquire a lock
+ * exactly once per device when calling `get_resources`. Subsequent calls
+ * will still be thread-safe but will not require a lock.
+ *
+ * All public methods of the `device_resources_manager` are static. Please see
+ * documentation of those methods for additional usage information.
+ *
+ */
+struct device_resources_manager {
+  device_resources_manager(device_resources_manager const&) = delete;
+  void operator=(device_resources_manager const&)           = delete;
+
+ private:
+  device_resources_manager() {}
+  ~device_resources_manager()
+  {
+    // Ensure that we destroy any pool memory resources before CUDA context is
+    // lost
+    per_device_components_.clear();
+  }
+
+  // Get an id used to identify this thread for the purposes of assigning
+  // (in round-robin fashion) the same resources to the thread on subsequent calls to
+  // `get_device_resources`
+  static auto get_thread_id()
+  {
+    static std::atomic<std::size_t> thread_counter{};
+    thread_local std::size_t id = ++thread_counter;
+    return id;
+  }
+
+  // This struct holds the various parameters used to control
+  // construction of the underlying resources shared by all
+  // `device_resources` objects returned by `get_device_resources`
+  struct resource_params {
```

Comment:
Every device will have the exact same set of `params` if I understand correctly?

Reply:
That is correct. My feeling on this was that if the application requires fine-grained control of these sorts of parameters per device, it probably is CUDA-aware enough to build its own `device_resources`. This still leaves the application in control of whether or not it uses a specific device, just not e.g. different number of total streams per device.